### PR TITLE
readers/nonforwarding: don't emit partition_end on next_partition,fast_forward_to

### PR DESCRIPTION
**Pull request by gusev-p**
The series fixes the `make_nonforwardable` reader, it shouldn't emit `partition_end` for previous partition after `next_partition()` and `fast_forward_to()`

Fixes: #12249

### CODE REVIEW

File: readers/mutation_reader.cc [hunk 1]
```diff
@@ -385,11 +385,6 @@ flat_mutation_reader_v2::~flat_mutation_reader_v2() {
     }
 }
 
-void flat_mutation_reader_v2::impl::forward_buffer_to(const position_in_partition& pos) {
-    clear_buffer();
-    _buffer_size = compute_buffer_size(*_schema, _buffer);
-}
```
Comment (denesb) [inline readers/mutation_reader.cc:391]: This smells like a bug to me. This method was supposed to clear buffer *up to* the specified position.
Did we drop this on the floor with the v2 migration? @tgrabiec?

File: readers/mutation_readers.cc [hunk 2]
```diff
@@ -411,15 +411,19 @@ flat_mutation_reader_v2 make_nonforwardable(flat_mutation_reader_v2 r, bool sing
         flat_mutation_reader_v2 _underlying;
         bool _single_partition;
         bool _static_row_done = false;
+        bool _partition_is_open = false;
```
Comment (denesb) [inline readers/mutation_readers.cc:414]: Can't we reuse the existing `_static_row_done` flag (maybe rename it to `_partition_header_done`)? The underlying forwarding reader will always emit the partition header (partition start and static row if any) first, then require fast forwarding to proceed.
I'm fine with 2 flags too if you think it helps code clarity.

File: test/boost/flat_mutation_reader_test.cc [hunk 3]
```diff
@@ -727,6 +728,33 @@ SEASTAR_TEST_CASE(test_make_forwardable) {
     });
 }
 
+SEASTAR_TEST_CASE(test_make_nonforwardable) {
```
Comment (denesb) [inline test/boost/flat_mutation_reader_test.cc:731]: We need test names that hint at what they are testing.

This test seems to stress the emitting logic for partition_end, so this should be in the name somehow.

**Review by denesb — CHANGES_REQUESTED** [review]
I don't understand the first commit at all. It seems to do some hacky workaround over a hack where fragments are injected into the middle of the reader stack. I think if this is causing problems, the injection should be fixed so that it happens on the bottom-most reader instead. Or maybe I completely misunderstood.

The rest looks good to me, with the exception of the last commit, which looks like some mistake made in the migration.

The PR title needs to be more specific: "fix the bug" is way too generic, it should be something like:
"readers/nonforwarding: don't emit partition_end on next_partition,fast_forward_to"

**Comment by gusev-p** [comment]
> I don't understand the first commit at all. It seems to do some hacky workaround over a hack where fragments are injected into the middle of the reader stack. I think if this is causing problems, the injection should be fixed so that it happens on the bottom-most reader instead. Or maybe I completely misunderstood.

Did you read the commit message?

**Comment by denesb** [comment]
> Did you read the commit message?

Yes, multiple times even.

**Comment by denesb** [comment]
> > Did you read the commit message?
> 
> Yes, multiple times even.

Looks like not enough times. I read it even and now I remembered that the cache reader does this trick where it pops off the partition start from the disk reader, because the content of the disk reader is injected into the middle of an already started partition. This is how the cache reader weaves together the cache and disk.
Let me have another look at the code too.

**Comment by gusev-p** [comment]
> I read it once again and now I remembered that the cache reader does this trick where it pops off the partition start from the disk reader

Yes, that's the point.

File: readers/mutation_readers.cc (hunk 2 above)
Comment (gusev-p) [reply readers/mutation_readers.cc:414]: > Can't we reuse the existing `_static_row_done` flag (maybe rename it to `_partition_header_done`)? The underlying forwarding reader will always emit the partition header (partition start and static row if any) first, then require fast forwarding to proceed.
> I'm fine with 2 flags too if you think it helps code clarity.

Reader should emit a `partition_end` only if there was a `partition_start` before. `_static_row_done` only tells us that the call to `fast_forward_to` was made, is says nothing about the content of the stream, it could be empty.

File: test/boost/flat_mutation_reader_test.cc (hunk 3 above)
Comment (gusev-p) [reply test/boost/flat_mutation_reader_test.cc:731]: This test is not only about `partition_end`, it contains all the basic checks for `make_nonforwardable`, there were no tests for this function before. Each check is a separate region with its own comment. This way it's easy to reuse various helpers and test data. We can separate each check into its own test, but I don't see much value in it. The already existing `test_make_forwardable` is structured in the same way.

Comment (denesb) [reply test/boost/flat_mutation_reader_test.cc:731]: We should add a test which subject this reader to the mutation source reader torture suite.

Comment (denesb) [reply test/boost/flat_mutation_reader_test.cc:731]: That suite would have cough all these problems

Comment (gusev-p) [reply test/boost/flat_mutation_reader_test.cc:731]: > We should add a test which subject this reader to the mutation source reader torture suite.

How to do it? Could you share an example?

**Comment by denesb** [comment]
Ok, the first commit makes sense now, but @tgrabiec should review too. Cache is tricky.

File: test/boost/flat_mutation_reader_test.cc (hunk 3 above)
Comment (denesb) [reply test/boost/flat_mutation_reader_test.cc:731]: Search for `run_mutation_source_tests` in `test/boost/mutation_reader_test.cc` or `test/boost/flat_mutation_reader_test.cc` and you will see plenty of examples.
The gist is this:
* You create an underlying reader with the passed-in mutations. The most convenient is to create a memtable, dump the mutations into them, then create a memtable reader. You will always have to create this with `streamed_mutation::forwarding::yes`.
* Wrap this reader with the nonforwarding reader.
* When the test asks for forwarding reader, you can use `make_forwarding()` to make your nonforwarding reader forwarding one.

**Comment by gusev-p** [comment]
The test failure seems related to https://github.com/scylladb/scylla-jmx/issues/194, not sure how to deal with it.

**Comment by denesb** [comment]
> The test failure seems related to [scylladb/scylla-jmx#194](https://github.com/scylladb/scylla-jmx/issues/194), not sure how to deal with it.

I don't see anything related to JMX in the test results. It looks like a timeout while trying to replace a node.

**Comment by gusev-p** [comment]
> > The test failure seems related to [scylladb/scylla-jmx#194](https://github.com/scylladb/scylla-jmx/issues/194), not sure how to deal with it.
> 
> I don't see anything related to JMX in the test results. It looks like a timeout while trying to replace a node.

Ah, I ran it locally, had some JMX issue, and though it was the same on the CI.

File: row_cache.cc [hunk 4]
```diff
@@ -637,12 +633,8 @@ class scanning_and_populating_reader final : public flat_mutation_reader_v2::imp
...
-        return _secondary_reader().then([this] (range_populating_reader::read_result&& res) {
-            auto&& [fropt, ps] = res;
+        return _secondary_reader().then([this] (flat_mutation_reader_v2_opt&& fropt) {
             if (fropt) {
-                if (ps) {
-                    push_mutation_fragment(std::move(*ps));
-                }
                 return make_ready_future<flat_mutation_reader_v2_opt>(std::move(fropt));
             } else {
                 _secondary_in_progress = false;
```
Comment (tgrabiec) [inline row_cache.cc:640]: > However, this makes a problem:
> now we pop partition_start from the underlying reader
> in autoupdating_underlying_reader::move_to_next_partition
> and manually push it back to downstream readers
> bypassing nonforwardable reader. This means if we
> change the logic in nonforwardable reader as described
> we will end up with partition_start without partition_end
> in the downstream readers.

I don't think make_nonforwardable() should rely on this assumption, that is that it is invoked on a stream which was never consumed from. It's enough if make_nonforwardable()::reader::next_partition() sets a flag which puts the reader at EOS without emitting anything.

File: test/boost/flat_mutation_reader_test.cc (hunk 3 above)
Comment (tgrabiec) [reply test/boost/flat_mutation_reader_test.cc:731]: This reader can't work with the general mutation_source tests, because it works only with a single partition. mutation source tests assume that the mutation source can accept multiple partitions.

Comment (tgrabiec) [reply test/boost/flat_mutation_reader_test.cc:731]: We could extract a battery of single-partition tests out of the general mutation source battery of tests.

File: readers/mutation_reader.cc (hunk 1 above)
Comment (tgrabiec) [reply readers/mutation_reader.cc:391]: It was needed with flat_mutation_reader which had range_tombstone fragments. They could be relevant to the next position_range on fast_forward_to() because they could hold information which extends past the current range. With flat_mutation_reader_v2 we no longer have such fragments, so mutation_fragment::relevant_for_range_assuming_after() was dropped and this is effectively clearing all. We just forgot to remove it. So I think this change is fine.

**Review by tgrabiec — CHANGES_REQUESTED** [review]

File: row_cache.cc (hunk 4 above)
Comment (tgrabiec) [reply row_cache.cc:640]: But the patch is good as a cleanup.

**Comment by tgrabiec** [comment]
The cover letter should mention "Fixes #..."

File: row_cache.cc (hunk 4 above)
Comment (gusev-p) [reply row_cache.cc:640]: > It's enough if make_nonforwardable()::reader::next_partition() sets a flag which puts the reader at EOS without emitting anything.

It's true for the case when `single_partition=true`, and this indeed should be fixed/tested. In general, there may be other partitions and we should handle that if `make_nonforwardable` pretends to be generic. In particular, we have to handle situations when this next partition is empty or not empty. In the first case we shouldn't emit `partition_end`, in the second case we should. The `_partition_is_open` flag was added to distinguish these two cases. Here we strive to preserve the semantics of intra-partition fast-forwarding mode described in [your comment](https://github.com/scylladb/scylladb/issues/12249#issuecomment-1439061442).

**Comment by gusev-p** [comment]
v2:
  * fix `next_partition` and `fast_forward_to` when `single_partition` is false;
  * use `SEASTAR_THREAD_TEST_CASE`;
  * add `test_make_nonforwardable_from_mutations_as_mutation_source` which exercises `make_nonforwardable` with the help of `run_mutation_source_tests`.

**Comment by gusev-p** [comment]
v3:
  * rebased on master

**Comment by gusev-p** [comment]
@denesb @tgrabiec ping

File: test/boost/flat_mutation_reader_test.cc [hunk 5]
```diff
@@ -727,6 +728,33 @@ SEASTAR_TEST_CASE(test_make_forwardable) {
...
+    return seastar::async([] {
```
Comment (denesb) [inline test/boost/flat_mutation_reader_test.cc]: Should use `SEASTAR_THREAD_TEST_CASE`.

File: test/boost/flat_mutation_reader_test.cc [hunk 6]
```diff
@@ -728,124 +728,122 @@ SEASTAR_TEST_CASE(test_make_forwardable) {
     });
 }
 
-SEASTAR_TEST_CASE(test_make_nonforwardable) {
-    return seastar::async([] {
-        simple_schema s;
-        tests::reader_concurrency_semaphore_wrapper semaphore;
-        const auto permit = semaphore.make_permit();
+SEASTAR_THREAD_TEST_CASE(test_make_nonforwardable) {
```
Comment (denesb) [inline test/boost/flat_mutation_reader_test.cc:731]: This should not be a separate commit, instead the change should be retroactively applied to the past commits.

File: test/boost/flat_mutation_reader_test.cc [hunk 7]
```diff
@@ -846,6 +846,33 @@ SEASTAR_THREAD_TEST_CASE(test_make_nonforwardable) {
...
+            tracing::trace_state_ptr,
+            streamed_mutation::forwarding fwd_sm,
+            mutation_reader::forwarding) mutable {
+            auto reader = make_flat_mutation_reader_from_mutations_v2(schema,
+                std::move(permit),
+                squash_mutations(muts),
+                range,
+                slice,
+                streamed_mutation::forwarding::yes);
+            reader = make_nonforwardable(std::move(reader), false);
```
Comment (denesb) [inline test/boost/flat_mutation_reader_test.cc:877]: Instead of `false` you should pass `muts.size() == 1` (after the squashing).

**Review by denesb — APPROVED** [review]
LGTM, only some minor comments.

**Review by denesb — APPROVED** [review]

File: test/boost/flat_mutation_reader_test.cc (hunk 3 above)
Comment (gusev-p) [reply test/boost/flat_mutation_reader_test.cc]: I added `test_make_nonforwardable_from_mutations_as_mutation_source`

File: test/boost/flat_mutation_reader_test.cc (hunk 7 above)
Comment (gusev-p) [reply test/boost/flat_mutation_reader_test.cc]: done

**Comment by gusev-p** [comment]
v4:
  * `test_make_nonforwardable_from_mutations_as_mutation_source` revealed a problem in `make_forwardable::reader::next_partition`, added a fix for it;
  * `flat_mutation_reader_test` cleanup, `seastar::async -> SEASTAR_THREAD_TEST_CASE`;
  * `test_make_nonforwardable_from_mutations_as_mutation_source`, added `single_partition = squashed_muts.size() == 1;` condition to check the logic when `single_partition==true`;
  * `make_nonforwardable::reader::next_partition` adjusted condition when `_single_partition==true`, set `EOS` only if `_partition_is_open==true`

**Comment by gusev-p** [comment]
@denesb ping, new version (hope the last one), tests passed

**Review by denesb — APPROVED** [review]

**Review by tgrabiec — APPROVED** [review]
